tools/proto-js-plugin/src/main/java/io/spine/js/generate/resolve/ExternalModule.java

```diff
@@ -0,0 +1,140 @@
+ */
+
+package io.spine.js.generate.resolve;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+import io.spine.code.js.DirectoryReference;
+import io.spine.code.js.FileReference;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.spine.util.Preconditions2.checkNotEmptyOrBlank;
+
+/**
+ * An external JavaScript module used in a project.
+ *
+ * <p>External means that it is provided by an artifact repository like NPM.
+ */
+public final class ExternalModule {
+
+    private final String name;
+    private final List<DirectoryPattern> directories;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param name
+     *         the name of the module
+     * @param directories
+     *         patterns of directories provided by the module
+     */
+    public ExternalModule(String name, List<DirectoryPattern> directories) {
+        this.name = checkNotEmptyOrBlank(name);
+        this.directories = ImmutableList.copyOf(directories);
+    }
+
+    /**
+     * Obtains the file reference within this module.
+     *
+     * @param fileReference
+     *         the relative file reference
+     * @return the file reference in this module
+     * @throws IllegalStateException
+     *         if the file is not provided by the module
+     */
+    FileReference fileInModule(FileReference fileReference) {
+        Optional<DirectoryPattern> matchingDirectory = matchingDirectory(fileReference);
+        checkState(matchingDirectory.isPresent());
+        DirectoryReference directory = matchingDirectory.get()
+                                                        .transform(fileReference.directory());
+        String fileName = fileReference.fileName();
+        String path = Joiner.on(FileReference.separator())
+                            .join(name, directory, fileName);
+        return FileReference.of(path);
+    }
+
+    /**
+     * Checks if the module provides the referenced file.
+     *
+     * @param fileReference
+     *         the file to check
+     * @return {@code true} if the module provides the file
+     */
+    boolean provides(FileReference fileReference) {
+        boolean result = matchingDirectory(fileReference).isPresent();
+        return result;
+    }
+
+    private Optional<DirectoryPattern> matchingDirectory(FileReference fileReference) {
+        DirectoryReference directory = fileReference.directory();
+        for (DirectoryPattern pattern : directories) {
+            if (pattern.matches(directory)) {
+                return Optional.of(pattern);
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * <a href="https://github.com/SpineEventEngine/web">The Spine Web</a> module.
+     */
+    public static ExternalModule spineWeb() {
+        List<DirectoryPattern> directories = ImmutableList.of(
```

Comment:
Not sure where we include/generate the JS files for Protobuf library types, such as `Any`.